javascripts/govuk/analytics/govuk-tracker.js

```diff
@@ -0,0 +1,115 @@
+;(function (global) {
+  'use strict'
+
+  var $ = global.jQuery
+  var GOVUK = global.GOVUK || {}
+
+  var GOVUKTracker = function (gifUrl, fieldsObject) {
+    this.gifUrl = gifUrl
+    this.dimensions = []
+    global.ga(function (tracker) {
+      this.gaClientId = tracker.get('clientId')
+    }.bind(this))
+  }
+
+  GOVUKTracker.load = function () {}
+
+  GOVUKTracker.prototype.trackPageview = function (path, title, options) {
+    var pageviewObject
+
+    if (typeof path === 'string') {
+      pageviewObject = { page: path }
+    }
+
+    if (typeof title === 'string') {
+      pageviewObject = pageviewObject || {}
+      pageviewObject.title = title
+    }
+
+    // Set an options object for the pageview (e.g. transport, sessionControl)
+    // https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#transport
+    if (typeof options === 'object') {
+      pageviewObject = $.extend(pageviewObject || {}, options)
+    }
+
+    if (!$.isEmptyObject(pageviewObject)) {
+      this.sendToTracker('pageview', pageviewObject)
+    } else {
+      this.sendToTracker('pageview')
+    }
+  }
+
+  // https://developers.google.com/analytics/devguides/collection/analyticsjs/events
+  GOVUKTracker.prototype.trackEvent = function (category, action, options) {
+    options = options || {}
+    var evt = {
+      eventCategory: category,
+      eventAction: action,
+      eventLabel: options.label
+    }
+
+    if (options.value) {
+      evt.eventValue = options.value.toString()
+    }
+
+    if (typeof options === 'object') {
+      $.extend(evt, options)
+    }
+
+    this.sendToTracker('event', evt)
+  }
+
+  GOVUKTracker.prototype.trackSocial = function (network, action, target) {
+    this.sendToTracker('social', {
+      'socialNetwork': network,
+      'socialAction': action,
+      'socialTarget': target
+    })
+  }
+
+  GOVUKTracker.prototype.addLinkedTrackerDomain = function () { /* noop */ }
+
+  GOVUKTracker.prototype.setDimension = function (index, value) {
+    this.dimensions['dimension_' + index] = value
+  }
+
+  GOVUKTracker.prototype.payloadParams = function (type, payload) {
+    var data = $.extend({},
+      payload,
+      this.dimensions,
+      {
+        eventType: type,
+        referrer: global.document.referrer,
+        gaClientId: this.gaClientId
+      }
+    )
+
+    if (global.performance) {
+      data.navigationType = global.performance.navigation.type.toString()
+      data.redirectCount = global.performance.navigation.redirectCount.toString()
+
+      for (var k in global.performance.timing) {
+        var v = global.performance.timing[k]
+        if (typeof v === 'string' || typeof v === 'number') {
+          data['timing_' + k] = v.toString()
+        }
+      }
+    }
+
+    return $.param(data)
+  }
+
+  GOVUKTracker.prototype.sendToTracker = function (type, payload) {
+    var doc = global.document
+    $(doc).on('readystatechange', function () {
```

Comment:
We need to wait until the `readyState` is `complete` so we can get more timings, e.g. `domInteractive` and `domComplete`